src/file.cr

```diff
@@ -282,13 +282,35 @@ class File < IO::FileDescriptor
   def self.extname(filename) : String
     filename.check_no_null_byte
 
-    dot_index = filename.rindex('.')
+    reader = Char::Reader.new(at_end: filename)
 
-    if dot_index && dot_index != filename.size - 1 && dot_index - 1 > (filename.rindex(SEPARATOR) || 0)
-      filename[dot_index, filename.size - dot_index]
-    else
-      ""
+    # if the pattern is foo. it has no extension
+    return "" if reader.current_char == '.'
+
+    # position the reader at the last . or SEPARATOR
+    while (current_char = reader.current_char) &&
+          (current_char != SEPARATOR && current_char != '.') &&
```

Comment:
Why not use `reader.current_char` here? THis current code makes me think that `reader.current_char` returning nil is an exit condition for the loop - it's not and `reader.current_char` cannot return nil.

Reply:
Actually, `Char::Reader` is not needed at all for this. You can simply use `string.to_slice` and go backwards from there, byte per byte. All of '.', `/` and `\` fit in a single byte in UTF-8 so they can't be mistaken for something else. And going byte per byte is both faster and simpler (no need to decode UTF-8).

This is also how it's done in Go.